sbi/neural_nets/estimators/categorical_net.py

```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
+        conditional density at a given time.
 
         Args:
-            condition: batch of context parameters for the net.
+            input: Inputs datapoints of shape `(batch_size, *input_shape)`
+            condition: Conditioning variable. `(batch_size, *condition_shape)`
 
         Returns:
-            Tensor: batch of predicted categorical probabilities.
+            Predicted categorical logits. `(batch_size, *input_shape,
+                num_categories)`
         """
-        # forward path
-        condition = self.activation(self.input_layer(condition))
+        embedded_condition = self.embedding_net.forward(condition)
+        out = super().forward(input, context=embedded_condition)
+        # masks out logits i.e. for variables with num_categories < max(num_categories)
+        return out.masked_fill(~self.mask.bool().flatten(), float("-inf"))
 
-        # iterate n hidden layers, input condition and calculate tanh activation
-        for layer in self.hidden_layers:
-            condition = self.activation(layer(condition))
-
-        return self.softmax(self.output_layer(condition))
-
-    def log_prob(self, input: Tensor, condition: Tensor) -> Tensor:
-        """Return categorical log probability of categories input, given condition.
+    def log_prob(self, input: Tensor, condition: Optional[Tensor] = None) -> Tensor:
+        r"""Return log-probability of samples.
 
         Args:
-            input: categories to evaluate.
-            condition: parameters.
+            input: Input datapoints of shape `(batch_size, *input_shape)`.
+            condition: Conditioning variable. `(batch_size, *condition_shape)`.
 
         Returns:
-            Tensor: log probs with shape (input.shape[0],)
+            Log-probabilities of shape `(batch_size,)`.
         """
-        # Predict categorical ps and evaluate.
-        ps = self.forward(condition)
-        # Squeeze the last dimension (event dim) because `Categorical` has
-        # `event_shape=()` but our data usually has an event_shape of `(1,)`.
-        return Categorical(probs=ps).log_prob(input.squeeze(dim=-1))
+        outputs = self.forward(input, condition=condition)
 
-    def sample(self, sample_shape: torch.Size, condition: Tensor) -> Tensor:
-        """Returns samples from categorical random variable with probs predicted from
-        the neural net.
+        outputs = outputs.reshape(*input.shape, self.num_categories)
+        log_prob = Categorical(logits=outputs).log_prob(input).sum(dim=-1)
 
-        Args:
-            sample_shape: number of samples to obtain.
-            condition: batch of parameters for prediction.
+        return log_prob
 
-        Returns:
-            Tensor: Samples with shape (num_samples, 1)
-        """
+    def sample(self, sample_shape, context=None):
+        # Ensure sample_shape is a tuple
+        if isinstance(sample_shape, int):
+            sample_shape = (sample_shape,)
```

Comment:
I think we should catch this with type hinting and it will raise an error otherwise.